File: backend/app/api/v1/regions.py
"""Region assembly for read routes.

A mapped answer that continues onto a later page owns regions on several rows; the UI
must receive all of them, so continuation children are folded in here.
"""
from __future__ import annotations

from app.db.models import Answer
from app.db.repositories import AnswerRegionRepository, AnswerRepository
from app.db.session import UnitOfWork
from app.schemas.common import Region


def answer_regions(
    session: UnitOfWork, organization_id: str, assessment_id: str
) -> dict[str, list[Region]]:
    """Regions per answer id, including every continuation segment of that answer."""
    rows: list[Answer] = AnswerRepository(session).for_assessment(organization_id, assessment_id)
    regions = AnswerRegionRepository(session).for_answers(organization_id, [row.id for row in rows])

    by_row: dict[str, list[Region]] = {}
    for region in regions:
        by_row.setdefault(region.answer_id, []).append(
            Region(page=region.page_number, bbox=region.bbox)
        )

    by_external = {row.external_id: row for row in rows}
    children: dict[str, list[Answer]] = {}
    for row in rows:
        if row.is_continuation_of and row.is_continuation_of in by_external:
            children.setdefault(by_external[row.is_continuation_of].id, []).append(row)

    out: dict[str, list[Region]] = {}
    for row in rows:
        collected = list(by_row.get(row.id, []))
        cursor = [row]
        while cursor:
            current = cursor.pop()
            for child in children.get(current.id, []):
                collected.extend(by_row.get(child.id, []))
                cursor.append(child)
        out[row.id] = sorted(collected, key=lambda region: (region.page, region.bbox.y1))
    return out

File: apps/web/lib/server/api/regions.ts
/**
 * Region assembly for read routes. Port of backend/app/api/v1/regions.py.
 *
 * A mapped answer that continues onto a later page owns regions on several rows;
 * the UI must receive all of them, so continuation children are folded in here.
 */
import type { UnitOfWork } from "@/lib/server/db/session";
import { AnswerRegionRepository, AnswerRepository } from "@/lib/server/db/repositories";
import type { Answer } from "@/lib/server/db/models";
import type { Region } from "@/lib/server/modules/common";

/** Regions per answer id, including every continuation segment of that answer. */
export async function answerRegions(
  session: UnitOfWork,
  organizationId: string,
  assessmentId: string
): Promise<Map<string, Region[]>> {
  const rows: Answer[] = await new AnswerRepository(session).forAssessment(organizationId, assessmentId);
  const regionRows = await new AnswerRegionRepository(session).forAnswers(organizationId, rows.map((r) => r.id));

  const byRow = new Map<string, Region[]>();
  for (const region of regionRows) {
    const list = byRow.get(region.answerId) ?? [];
    list.push({
      page: region.pageNumber,
      bbox: { x1: region.bbox[0], y1: region.bbox[1], x2: region.bbox[2], y2: region.bbox[3] },
    });
    byRow.set(region.answerId, list);
  }

  const byExternal = new Map(rows.map((row) => [row.externalId, row]));
  const children = new Map<string, Answer[]>();
  for (const row of rows) {
    if (row.isContinuationOf && byExternal.has(row.isContinuationOf)) {
      const parent = byExternal.get(row.isContinuationOf)!;
      const list = children.get(parent.id) ?? [];
      list.push(row);
      children.set(parent.id, list);
    }
  }

  const out = new Map<string, Region[]>();
  for (const row of rows) {
    const collected = [...(byRow.get(row.id) ?? [])];
    const cursor: Answer[] = [row];
    while (cursor.length > 0) {
      const current = cursor.pop()!;
      for (const child of children.get(current.id) ?? []) {
        collected.push(...(byRow.get(child.id) ?? []));
        cursor.push(child);
      }
    }
    collected.sort((a, b) => (a.page !== b.page ? a.page - b.page : a.bbox.y1 - b.bbox.y1));
    out.set(row.id, collected);
  }
  return out;
}
